[Chin Zhe Ning] iP

chinzhening opened the pull request:
### Mike
> "You and I are a team." – Mike Wazowski ([wiki](https://en.wikipedia.org/wiki/Mike_Wazowski))

Mike frees your mind of having to keep track of things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/biinnnggggg/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

***

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
      Application.launch(MainApp.class, args);
   }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

chinzhening committed:
Chatbot renamed from 'Duke' to 'Mike Wazowski', and now greets the user.

chinzhening committed:
Added echo as the default behaviour.

chinzhening committed:
Added Task and TaskList classes, default behaviour is to add a task.

chinzhening committed:
Added support for marking tasks as completed.

chinzhening committed:
Added Event, Todo, and Deadline. Added some minimal error handling. Increased the parsing capability.

chinzhening committed:
1. Changed the logo of Mike to be compatible with windows-1252 encoding to be in line with the automated text ui testing.
2. Added commands to input.txt
2. Improved comprehensiveness of the input error handling. Error messages are now more helpful.

chinzhening committed:
Refactored parsing code into a CommandParserClass and command execution from parsing

chinzhening committed:
Added response method to AddCommand. Added support for deleting tasks from the list.

chinzhening committed:
Separated scanning from parsing. CommandScanner outputs a list of tokens that the parser parses. Enum class is used for the token types.

chinzhening committed:
no message

chinzhening committed:
Revert "no message"

This reverts commit 66a3454da5d87c74b95a85b9e97d4923a5edc476.

chinzhening committed:
Renamed fields in Deadline and Event related classes.

chinzhening committed:
Added getters.

chinzhening committed:
Refactored class vocab to be more descriptive.

chinzhening committed:
Todo

chinzhening committed:
Added method header for javadocs.

chinzhening committed:
Format switch statement according to CS2103 iP coding standard

chinzhening committed:
`Add support for load and save.

chinzhening committed:
Refactor Ui methods to static.

chinzhening committed:
Update .gitignore to ignore /data/

chinzhening committed:
Remove /data/ from repo.

chinzhening committed:
Add date

chinzhening committed:
Refactor load save functionality in TaskList to Storage object

chinzhening committed:
Refactor into mike, mike.task, and mike.command packages

chinzhening committed:
Add support for different views of the same list.

This is essentially filter capability for listing tasks. Currently only filtering by date is supported.

chinzhening committed:
Merge remote-tracking branch 'origin/add-gradle-support'

chinzhening committed:
Add JavaDoc comments.

chinzhening committed:
Refactor code to adhere to coding standard.

chinzhening committed:
Fix list view bug and add find command.

chinzhening committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/mike/CommandParser.java
#	src/main/java/mike/CommandScanner.java
#	src/main/java/mike/command/Command.java
#	src/main/java/mike/command/ExitCommand.java

chinzhening committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/mike/CommandParser.java
#	src/main/java/mike/command/AddCommand.java
#	src/main/java/mike/command/Command.java
#	src/main/java/mike/task/Event.java
#	src/main/java/mike/task/Task.java

chinzhening committed:
Add more JavaDocs.

chinzhening committed:
Refactor CommandScanner methods to have return value instead of void.

chinzhening committed:
Refactor execute methods in the Command classes to return a String response. These responses are then displayed by Ui in Mike.

chinzhening committed:
Add testing for CommandParser and CommandScanner.

chinzhening committed:
Add support for producing jar files

vnnamng commented on src/main/java/mike/Mike.java:
```diff
@@ -0,0 +1,64 @@
+        ui.displayWelcome();
+
+        boolean exitSeen = false;
+
```
Should this boolean flag be renamed?

vnnamng commented on src/main/java/mike/Storage.java:
```diff
@@ -0,0 +1,118 @@
+            throw new MikeException("File corrupted. Unable to load saved data.");
+        }
+
+        if (taskIsCompleted) {
```
Name make sense but coding standard dont like...

chinzhening replied:
Nice catch, it will be changed.

vnnamng commented on src/main/java/mike/TaskList.java:
```diff
@@ -0,0 +1,24 @@
+ * TaskList is a wrapper class for an ArrayList of {@link Task} objects.
+ * @author ningc
+ */
+public class TaskList extends ArrayList<Task>  {
+    public String view(ListView listView) {
+        return this
+                .stream()
+                .filter(task -> task.inListView(listView))
+                .map(task -> {
+                    int index = this.indexOf(task) + 1;
+                    return index + "." + task;
+                })
+                .reduce("", (s1, s2) -> s1 + "\n" + s2);
+    }
+
```
nice Stream bro

chinzhening replied:
Thanks! I am quite proud of it 😂😂

vnnamng commented on src/main/java/mike/Ui.java:
```diff
@@ -0,0 +1,103 @@
+     * Prints the Wazowski logo as ASCII art.
+     */
+    private void displayLogo() {
+        String logo =
+                "                   .\n" +
+                "          &#  #*&/,,(@%*. (#\n" +
+                "           %#((,.      #%,*,\n" +
+                "         %%%(/,.. *(     /,.**\n" +
+                "      .%%%%#/**,%(@@&@    /.,,/\n" +
+                "      %%%%#(((/*(#*(%&    ..,,,*(\n" +
+                "    *#%%%##(/,#(/**,,...,...,,,**/\n" +
+                "   .#%%%%%##(((//#(//*......,,,**//\n" +
+                "   &#%%%%%%###(///*,,..,,,,,,,***/#*\n" +
+                "   &#%%%%%%#####@@@@@@(,,**,****/(#*\n" +
+                "  *&%#%%&&&%%%@@@@@@@@@@&**/////(.&/*\n" +
+                "  #&, #%&&&&&&%(((/*****((((((##.  #(\n" +
+                "  ##    #&&&&&&&&&&%%%%%%###%%(    (*\n" +
+                "  .#/     /%%%%%%%%%%%%%%%%#       //\n" +
+                "   #(     /##            *#,       (/\n" +
+                "   %*     /#(             #,       //\n" +
+                "   %(/.    %(            .%&      //*/\n" +
+                "   %/,&,   &(            #%/      (#//,\n" +
+                "   /(*     ##*           #%       &*#%\n" +
+                "     @/@*  (#*.          #(.      /%%\n" +
+                "      (/(/##(...........##(//&#(&/.\n" +
+                "     (@%(&@@,#,..........,*@@&&@&*,\n";
+        display(logo);
+    }
```
This logo should be saved as a static constant

vnnamng commented on src/main/java/mike/Ui.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private void displayGreeting() {
+        String greeting =
+                " Hello! I'm mike WAZOWSKI.\n" +
+                " What can I do for you?";
+        display(greeting);
+    }
+
+    /**
+     * Displays a horizontal line to the user.
+     */
+    public static void displayLine() {
+        String horizontalLine = "==========================================";
+        display(horizontalLine);
+    }
```
static constant same as the above, minor stuff no fetch

vnnamng commented on src/main/java/mike/CommandParser.java:
```diff
@@ -0,0 +1,293 @@
+        return new MikeException(message);
+    }
+
+    private Token consume(TokenType type, String message) throws MikeException {
+        if (check(type)) {
+            return advance();
+        }
+        throw error(message);
+    }
+
+    private boolean check(TokenType type) {
+        return peekToken().getType() == type;
+    }
+
+    private Token advance() {
+        if (!isAtEnd()) {
+            current++;
+        }
+        return previousToken();
+    }
+
+    private Token previousToken() {
+        return tokens.get(current - 1);
+    }
+
+    private boolean isAtEnd() {
+        return peekToken().getType() == TokenType.EOC;
+    }
+
+    private Token peekToken() {
+        return tokens.get(current);
+    }
```
the method names specific enough

chinzhening replied:
My reasoning was that these methods are private and only meant for use inside the class so it would be specific enough in the context. What do you think?

vnnamng reviewed (commented):
good work bro!

chinzhening committed:
Add checkstyle support

chinzhening committed:
Fix coding standard violations

chinzhening committed:
Improve code quality

chinzhening committed:
Add JavaFx GUI

chinzhening committed:
Fix exit and save functionality.

chinzhening committed:
Style Gui and remove send button.

chinzhening committed:
Update build Gradle file

Build Gradle file has some outdated field attributes from previous
versions, e.g., `mainClassName`.

The outdated attribute `mainClassName` cause the entry point of the
application to incorrect.

Updating the attribute to the correct entry point `Launcher` will
fix this issue.

chinzhening committed:
Improve Javadoc comment for the view method in `TaskList`

The Javadoc comment for view was not descriptive enough.

Editing the Javadoc comment to be more descriptive is a small step in
improving the overall code quality of the project. This will be
beneficial for future developments.

Let us change 'Filters' to 'Returns' and improve the grammar of the
sentence.

It is customary for the Javadoc comment of return statements to start
with the word 'Returns', a verb.

chinzhening committed:
Add Java assert statements.

In class `Main` we have some important assumptions that are not
validated:
* fxmlLoader is not null
* ap is not null
without which the program will crash.

Adding assert statements into `Main` will validate these assumptions
and let us know when these assumptions are violated.

Let us add two assert statements, one after the FXMLLoader constructor
is called, and another after the AnchorPane constructor is called.

Using assert statements is preferable because it will fulfill some
of the project requirements mandated by CS2103.

Refer to the [project deliverables page]
(https://nus-cs2103-ay2324s2.github.io/website/schedule/week5/project.html)

chinzhening committed:
Refactor 'CommandParser' class

'CommandParser' class had very long methods. Methods
contained multiple layers of abstraction.

The common behaviors across parse methods cause code duplication.
Long methods made the code less readable. The multiple layers of
abstraction made it harder to understand and follow.

Extracting the common behavior into more atomic private methods allows
us to reduce code duplication and at the same time enhance readability.

This improves the overall code quality of the class.

chinzhening committed:
Change `CommandScanner` method names

Change the method names to be more intuitive. For example,
`getStartToken` is renamed to `scanStartToken`. This helps improve the
code quality.

chinzhening committed:
Refactor the `Storage` class method `load`

Multiple layers of abstraction present in `load`.

Extracting out grouped behavior to private methods will enhance
readability of code and improve the code quality.

chinzhening committed:
Merge pull request #2 from biinnnggggg/branch-A-CodeQuality

Branch a code quality

chinzhening committed:
Merge branch 'master' into branch-A-Assertions

chinzhening committed:
Merge branch 'master' into branch-A-Streams

chinzhening committed:
Merge pull request #4 from biinnnggggg/branch-A-Assertions

Add Java assert statements.

chinzhening committed:
Merge branch 'master' into branch-A-Streams

chinzhening committed:
Merge pull request #3 from biinnnggggg/branch-A-Streams

Improve Javadoc comment for the view method in `TaskList`

chinzhening committed:
Add find command functionality

Find command now has the optional parameter to enable fuzzy search:
`find [keyword] /fuzzy [on/off]`
This will be a great quality of life improvement for most users.

Changed `CommandParser` to be able to parse the new command format.
Added bitap algorithm to `ListView` where fuzzy search is performed.
`Deadline` and `Task` now adopt the new constructors for `ListView`.

chinzhening committed:
Add ARCHIVE as a TokenType and "archive" as a keyword in
`CommandScanner`

chinzhening committed:
Update `execute` method signature to take `Storage` object as a
parameter.

Method implementation does not have a reference to `Storage` object and
therefore cannot act on it. This prevents archive functionality.

chinzhening committed:
Add support for `ArchiveCommand`

To do this, we
* create parsing for `ArchiveCommand` in the appropriate
class `CommandParser`.
* add archive file creation functionality in `Storage` to be used in
`ArchiveCommand` execution.

chinzhening committed:
Fix parameter parsing bug

chinzhening committed:
Attribute fuzzy search algorithm to comply with reuse policies

chinzhening committed:
Disable Gui window resizing

chinzhening committed:
Fix coding standard violations in `CommandParser`, `ListView`, and
`FindCommand`

chinzhening committed:
Add representative screenshot of the Ui to docs

chinzhening committed:
Add user guide to explain all the important features of the chatbot

chinzhening committed:
Format responses to mark, unmark, event commands

Responses were not consistent with the ones for todo and deadline.

Change the responses to be consistent.

chinzhening committed:
Remove logo and update README for the user guide
